onmt/Models.py

```diff
@@ -425,6 +429,117 @@ def forward(self, src, tgt, lengths, dec_state=None):
             attns = None
         return out, attns, dec_state
 
+    @property
+    def lr(self):
+        return self.optimizer.param_groups[0]['lr']
+
+    @lr.setter
+    def lr(self, lr):
+        self.optimizer.param_groups[0]['lr'] = lr
+
+    def setup_optimizer(self, optimizer, initial_lr, max_grad_norm):
+        if not isinstance(optimizer, six.string_types):
+            # train_from case
+            self.optimizer = optimizer
+        elif optimizer == 'sgd':
+            self.optimizer = optim.SGD(self.parameters(), lr=initial_lr)
+        elif optimizer == 'adagrad':
+            self.optimizer = optim.Adagrad(self.parameters(), lr=initial_lr)
+        elif optimizer == 'adadelta':
+            self.optimizer = optim.Adadelta(self.parameters(), lr=initial_lr)
+        elif optimizer == 'adam':
+            self.optimizer = optim.Adam(self.parameters(), lr=initial_lr,
+                                        betas=[0.9, 0.98], eps=1e-9)
+        else:
+            raise RuntimeError("Invalid optim method: " + self.method)
+
+        self.max_grad_norm = max_grad_norm
+
+    def setup_lr_scheduler(self, opt, checkpoint=None):
+        assert self.optimizer is not None
+
+        if checkpoint is not None:
+            last_epoch = checkpoint['epoch']
+            last_lr_steps = self.optimizer.state['last_lr_steps']
+        else:
+            last_epoch = 0  # for epoch-level lr update.
+            last_lr_steps = 0  # for batch-level lr update.
+
+        # LambdaLR requires 'initial_lr' set if last_epoch is passed in.
+        self.optimizer.param_groups[0]['initial_lr'] = self.lr
+
+        # 1. Batch level lr update: using 'noam' method.
+        self.lr_batch_scheduler = None
```

Comment:
Do these all need to be member variables? The aim of the PR was to have less state associated with Optimization.